predicators/envs/doors.py

```diff
@@ -729,3 +729,447 @@ def _get_open_door_target_value(mass: float, friction: float,
+                                         [self._door_type, self._door_type],
+                                         self._DoorsShareRoom_holds)
+        # Static objects (always exist no matter the settings).
+        self._robot = Object("robby", self._robot_type)
+        # Hyperparameters from CFG.
+        self._room_map_size = CFG.doors_room_map_size
+        self._min_obstacles_per_room = CFG.doors_min_obstacles_per_room
+        self._max_obstacles_per_room = CFG.doors_max_obstacles_per_room
+        self._min_room_exists_frac = CFG.doors_min_room_exists_frac
+        self._max_room_exists_frac = CFG.doors_max_room_exists_frac
+        # See note in _sample_initial_state_from_map().
+        self._task_id_count = itertools.count()
+        self._door_to_knob: Dict[Object, Object] = {}
+        self._open_door_target_value = CFG.doorknobs_target_value
+
+    @classmethod
+    def get_name(cls) -> str:
+        return "doorknobs"
+
+    def simulate(self, state: State, action: Action) -> State:
+        assert self.action_space.contains(action.arr)
+        dx, dy, new_door_rot = action.arr
+        x = state.get(self._robot, "x")
+        y = state.get(self._robot, "y")
+        new_x = x + dx
+        new_y = y + dy
+        next_state = state.copy()
+        next_state.set(self._robot, "x", new_x)
+        next_state.set(self._robot, "y", new_y)
+        # Check for collisions.
+        if self.state_has_collision(next_state):
+            # Revert the change to the robot position.
+            next_state.set(self._robot, "x", x)
+            next_state.set(self._robot, "y", y)
+        # If touching a door, change its value based on the action.
+        for door in state.get_objects(self._door_type):
+
+            if self._TouchingDoor_holds(state, [self._robot, door]):
+                # Get corresponding doorknob and rotate it accordingly
+                doorknob = self._door_to_knob[door]
+                new_door_level = np.clip(
+                    state.get(doorknob, "rot") + new_door_rot, 0.0, 1.0)
+                next_state.set(doorknob, "rot", new_door_level)
+                # Check if we should now open the door.
+                target = self._open_door_target_value
+                if abs(new_door_level - target) < self.open_door_thresh:
+                    next_state.set(door, "open", 1.0)
+                    next_state.set(doorknob, "open", 1.0)
+                else:
+                    next_state.set(door, "open", 0.0)
+        return next_state
+
+    def _generate_train_tasks(self) -> List[EnvironmentTask]:
+        self._room_map_size = CFG.doors_room_map_size
+        return self._get_tasks(num=CFG.num_train_tasks, rng=self._train_rng)
+
+    def _generate_test_tasks(self) -> List[EnvironmentTask]:
+        self._room_map_size = CFG.test_doors_room_map_size
+        return self._get_tasks(num=CFG.num_test_tasks, rng=self._test_rng)
+
+    @property
```

Comment:
in general with this and below, it looks like there's a lot of duplicated code from the superclass. Could you cut down on this and make sure you only override things that need to be changed? I think that'll make review + future code management a lot easier.